File: XDownloader/app/src/main/java/com/pv/xdownloader/ui/tour/OverflowViewPager.java
package com.pv.xdownloader.ui.tour;

import android.content.Context;
import android.support.v4.view.ViewPager;
import android.util.AttributeSet;

public class OverflowViewPager extends ViewPager {
    float startDragX;
    OnSwipeOutListener listener;
    private boolean isLastPageSwiped;
    private int counterPageScroll;

    public OverflowViewPager(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public void setListener(OnSwipeOutListener listener) {
        this.listener = listener;
    }

    @Override
    public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
        super.onPageScrolled(position, positionOffset, positionOffsetPixels);
        int lastIndex = getAdapter().getCount() - 1;
        boolean isLastPosition = position == lastIndex;
        if (isLastPosition && positionOffset == 0 && !isLastPageSwiped) {
            if (counterPageScroll != 0) {
                isLastPageSwiped = true;
                if (listener != null) {
                    listener.onSwipeOutAtEnd();
                }
            }
            counterPageScroll++;
        } else {
            counterPageScroll = 0;
        }
    }


    public interface OnSwipeOutListener {

        public void onSwipeOutAtEnd();

    }
}

File: XDownloader/app/src/main/java/com/pv/xdownloader/ui/guide/OverflowViewPager.kt
package com.pv.xdownloader.ui.guide

import android.content.Context
import android.support.v4.view.ViewPager
import android.util.AttributeSet

class OverflowViewPager(context: Context, attrs: AttributeSet) : ViewPager(context, attrs) {
    internal var startDragX: Float = 0.toFloat()
    internal var listener: OnSwipeOutListener? = null
    private var isLastPageSwiped: Boolean = false
    private var counterPageScroll: Int = 0

    fun setListener(listener: OnSwipeOutListener) {
        this.listener = listener
    }

    public override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
        super.onPageScrolled(position, positionOffset, positionOffsetPixels)
        val lastIndex = adapter.count - 1
        val isLastPosition = position == lastIndex
        if (isLastPosition && positionOffset == 0f && !isLastPageSwiped) {
            if (counterPageScroll != 0) {
                isLastPageSwiped = true
                if (listener != null) {
                    listener!!.onSwipeOutAtEnd()
                }
            }
            counterPageScroll++
        } else {
            counterPageScroll = 0
        }
    }


    interface OnSwipeOutListener {

        fun onSwipeOutAtEnd()

    }
}
